### ISSUE TITLE
WS + Proxy + SSL

### ISSUE THREAD

**Issue by vboikotriller** [ninenines/gun#267]
Hello
What is the correct use of gun:connect ?
The following code raises connection timeout error

```
            case gun:open(ProxyHost, ProxyPort, #{transport => tls, protocols => [http]}) of
                {ok, ConnPid} ->
                    case gun:await_up(ConnPid) of
                        {ok, http} ->
                            _StreamRef = gun:connect(ConnPid, #{
                                host => Host,
                                port => Port
                            }),
                            StreamRef = gun:ws_upgrade(ConnPid, ?API_ENDPOINT, [
                                {?SESSION_HEADER, SecretToken},
                                {?DEVICE_ID_HEADER, DeviceId},
                                {?NAMESPACE, Namespace}]),
                            case wait_for_ws_upgrade(ConnPid, StreamRef, 5000) of
                                {error, Reason} ->
                                    gun:close(ConnPid),
                                    {error, Reason};
                                _ -> {ok, ConnPid, StreamRef}
                            end;
                        {error, Reason} ->
                            gun:close(ConnPid),
                            {error, Reason}
                    end;
                {error, Reason} -> {error, Reason}
            end;

```

**Comment by essen** [comment]
Lots of examples here: https://github.com/ninenines/gun/blob/master/test/tunnel_SUITE.erl

**Comment by vboikotriller** [comment]
I checked examples and I found they can't work with any proxy.
For example, I tried tinyproxy and it returns HTTP/1.0, which `gun` do not seem to support.
I tried squid, but proxy replies to CONNECT request only after request to URL.

I have not found a working configuration yet. Any hints ?
Current code:

```
wait_for_ws_upgrade(ConnPid, StreamRef, Timeout) ->
    io:format("waiting for WS upgrade~n"),
    receive
        Res1 ->
            io:format("received message: ~p~n", [Res1]),
            case Res1 of
                {gun_upgrade, ConnPid, StreamRef, [<<"websocket">>], _} ->
                    ResponseHeader = binary:encode_unsigned(?PT_INFO_REQUEST),
                    Packet = << 0, ResponseHeader/binary, <<0:(19 * 8)>>/binary, 1, 0, 0 >>,
                    gun:ws_send(ConnPid, StreamRef, {binary, Packet});
                {gun_response, ConnPid, _, _, Status, _Headers} ->
                    {error, io_lib:format("Status: ~p", [Status])};
                {gun_error, ConnPid, _StreamRef, Reason} ->
                    {error, io_lib:format("gun_error: ~p", [Reason])}
            end
    after Timeout -> {error, timeout}
    end.

tst() ->
    application:ensure_all_started(gun),
    Host = "something.something",
    Port = 443,
    ProxyHost = "127.0.0.1",
    ProxyPort = 3128,

    case inet:getaddr(Host, inet) of
        {ok, _IPv4} ->
            %% Initiate SSL connection
            case gun:open(ProxyHost, ProxyPort, #{transport => tcp, protocols => [http]}) of
                {ok, ConnPid} ->
                    io:format("connected to proxy~n"),
                    %% now creating connection through the proxy to origin host
                    StreamRef1 = gun:connect(ConnPid, #{
                                                        host => Host,
                                                        port => Port,
                                                        transport => tls,
                                                        protocols => [http]
                                                       }),
                    io:format("connected to host~n"),
                    case gun:await_up(ConnPid) of
                        {ok, http} ->
                            io:format("up!~n"),
                            timer:sleep(300), %% attempt to receive reply to CONNECT statement
                            StreamRef2 = gun:ws_upgrade(ConnPid, ?API_ENDPOINT,
                                                        [{?SESSION_HEADER, SecretToken}]),
                            io:format("WS upgrade sent~n"),
                            case wait_for_ws_upgrade(ConnPid, StreamRef2, 5000) of
                                {error, Reason} ->
                                    timer:sleep(10000),
                                    gun:close(ConnPid),
                                    {error, Reason};
                                _ -> {ok, ConnPid, StreamRef2}
                            end;
                        {error, Reason} ->
                            gun:close(ConnPid),
                            {error, Reason}
                    end;
                {error, Reason} -> {error, Reason}
            end
    end.


```

**Comment by essen** [comment]
At this point I have implemented the HTTP/1.1 and HTTP/2 specs. I have not tested against specific proxies although I did try against random proxies found online (proxy lists).

That said, I can see on your code that you do not provide the `tunnel` option. It may or may not make a difference for tinyproxy.

It sounds like squid works though, after you connect you will receive both an up and the response. After that you can use the tunnel. You are not receiving the CONNECT response currently (well, you do when you wait for the Websocket one).

**Comment by vboikotriller** [comment]
I noticed in documentation:

```
transport (tcp)
Transport that will be used for tunneled requests. Note that due to Erlang/OTP limitations it is not possible to tunnel a TLS connection inside a TLS tunnel.
```

But when I change `transport` from `tcp` to `tls`, I receive timeout error.
```

gun:open(ProxyHost, ProxyPort, #{transport => tls, protocols => [http]})
gun:await_up(ConnPid)
gun:connect(ConnPid, #{ host => Host, port => Port })
gun:ws_upgrade(ConnPid, ..)
..


```

**Comment by vboikotriller** [comment]
I checked this [file](https://github.com/ninenines/gun/blob/master/test/tunnel_SUITE.erl#L909)
and `gun` source code, but I have not found how to use `tunnel` option, as when I use `gun:connect`, I do not have StreamRef yet.
```
gun:open(ProxyHost, ProxyPort, #{transport => tls, protocols => [http]})
gun:await_up(ConnPid)
StreamRef = gun:connect(ConnPid, #{ host => Host, port => Port })
gun:ws_upgrade(ConnPid, ..)

```

**Comment by essen** [comment]
Tunnel option is given to ws_upgrade.

**Comment by vboikotriller** [comment]
`ws_upgrade` do not accept `tunnel` in options.
`gun:ws_upgrade(ConnPid, "/api", [], >#{tunnel => StreamRef}).
`
Error:
```
                    {error,
                        {options,{ws,{tunnel,#Ref<0.2968715942.236191745.144545>}}}}
     in function  gun:ws_upgrade/4 (src/gun.erl, line 629)

```

**Comment by essen** [comment]
It does, see https://github.com/ninenines/gun/blob/master/src/gun.erl#L927-L929

It appears that you are trying to do this using Gun 1.3 though. It did not have the tunnel option yet I think. I am not sure it works well with the option. It was not needed for the few cases that Gun 1.3 worked with.

Anyway I've already given you the solution here:

> It sounds like squid works though, after you connect you will receive both an up and the response. After that you can use the tunnel. You are not receiving the CONNECT response currently (well, you do when you wait for the Websocket one).

You need to receive the CONNECT response before you do the Websocket upgrade.

**Comment by imgrey** [comment]
Erlang variant for the history ( I spent about one week on that ).

```
    case inet:getaddr(Host, inet) of
        {ok, _IPv4} ->
            %% Initiate SSL connection
            case gun:open(ConnectHost, ConnectPort, #{transport => Transport, protocols => [http]}) of
                {ok, ConnPid} ->
                    case gun:await_up(ConnPid) of
                        {ok, http} ->
                            case maybe_proxy_tunnel(ProxyEnabled, ConnPid, Host, Port, Headers) of
                                {error, Reason} ->
                                    gun:close(ConnPid),
                                    {error, Reason};
                                StreamRef ->
                                    case wait_for_ws_upgrade(ConnPid, StreamRef, 5000) of
                                        {error, Reason} ->
                                            gun:close(ConnPid),
                                            {error, Reason};
                                        _ -> {ok, ConnPid, StreamRef}
                                    end
                            end;
                        {error, Reason} ->
                            gun:close(ConnPid),
                            {error, Reason}
                    end;
                {error, Reason} -> {error, Reason}
            end;

%%
%% Establishes proxy tunnel before upgrading http connection to websocket.
%%
maybe_proxy_tunnel(ProxyEnabled, ConnPid, Host, Port, Headers) when ProxyEnabled =:= true ->
    StreamRef0 = gun:connect(ConnPid, #{
                                        host => Host,
                                        port => Port,
                                        transport => tls}),
    case gun:await(ConnPid, StreamRef0) of
        {response, fin, 200, _} ->
            gun:ws_upgrade(ConnPid, ?API_ENDPOINT, Headers, #{tunnel => StreamRef0});
        {error, Reason} -> {error, Reason}
    end;
maybe_proxy_tunnel(ProxyEnabled, ConnPid, _, _, Headers) when ProxyEnabled =:= false ->
    gun:ws_upgrade(ConnPid, ?API_ENDPOINT, Headers).


wait_for_ws_upgrade(ConnPid, StreamRef, Timeout) ->
    receive
        {gun_tunnel_up, ConnPid, _TunnelStreamRef, _} ->
            wait_for_ws_upgrade(ConnPid, StreamRef, Timeout);
        {gun_upgrade, ConnPid, StreamRef, [<<"websocket">>], _} ->
            %% Respond to request
            ok;;
        {gun_response, ConnPid, _, _, Status, _Headers} ->
            {error, io_lib:format("Status: ~p", [Status])};
        {gun_error, ConnPid, _StreamRef, Reason} ->
            {error, io_lib:format("gun_error: ~p", [Reason])}
    after Timeout -> {error, timeout}
    end.

```
